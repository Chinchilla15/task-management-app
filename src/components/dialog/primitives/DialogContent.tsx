import { forwardRef, useEffect, useRef } from "react";
import { createPortal } from "react-dom";
import { cn } from "@/lib/utils";
import { useFocusTrap } from "@/hooks/useFocusTrap";
import { useDialogContext } from "../context/DialogContext";
import type { DialogContentProps } from "@types";

export const DialogContent = forwardRef<HTMLDivElement, DialogContentProps>(
  ({ children, className, ...props }, _ref) => {
    const { open, onOpenChange } = useDialogContext();
    const focusTrapRef = useFocusTrap(open, () => onOpenChange(false));
    const previousFocus = useRef<HTMLElement | null>(null);

    const handleBackdropClick = () => {
      onOpenChange(false);
    };

    useEffect(() => {
      if (open) {
        previousFocus.current = document.activeElement as HTMLElement;
        document.body.style.overflow = "hidden";
      } else if (previousFocus.current) {
        previousFocus.current.focus();
        document.body.style.overflow = "";
      }
      return () => {
        document.body.style.overflow = "";
      };
    }, [open]);

    if (!open) return null;

    return createPortal(
      <div className="fixed inset-0 z-50" role="presentation">
        <div
          className={cn("fixed inset-0 backdrop-blur-sm")}
          onClick={handleBackdropClick}
          aria-hidden="true"
        />
        <div
          ref={focusTrapRef}
          role="dialog"
          aria-modal="true"
          aria-labelledby="dialog-title"
          aria-describedby="dialog-description"
          className={cn(
            "fixed left-1/2 top-1/2 z-50 flex w-11/12 max-w-xl translate-x-[-50%] translate-y-[-50%] gap-4",
            "rounded-lg bg-neutral-3 p-4 shadow-lg",
            className,
          )}
          onClick={(e) => e.stopPropagation()}
          {...props}
        >
          {children}
        </div>
      </div>,
      document.body,
    );
  },
);

DialogContent.displayName = "DialogContent";
