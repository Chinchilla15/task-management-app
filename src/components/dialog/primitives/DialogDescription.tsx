import { forwardRef, HTMLAttributes } from "react";
import { cn } from "@/lib/utils";

export const DialogDescription = forwardRef<
  HTMLParagraphElement,
  HTMLAttributes<HTMLParagraphElement>
>(({ className, ...props }, ref) => (
  <p
    ref={ref}
    className={cn("text-muted-foreground text-sm", className)}
    {...props}
  />
));

DialogDescription.displayName = "DialogDescription";
