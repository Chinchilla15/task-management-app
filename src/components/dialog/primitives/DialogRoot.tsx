import { DialogContext } from "../context/DialogContext";
import type { DialogRootProps } from "@types";

export const DialogRoot = ({
  children,
  open,
  onOpenChange,
}: DialogRootProps) => {
  return (
    <DialogContext.Provider value={{ open, onOpenChange }}>
      {children}
    </DialogContext.Provider>
  );
};
