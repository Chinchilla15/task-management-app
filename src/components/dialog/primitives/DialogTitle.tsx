import { forwardRef, HTMLAttributes } from "react";
import { cn } from "@/lib/utils";

export const DialogTitle = forwardRef<
  HTMLHeadingElement,
  HTMLAttributes<HTMLHeadingElement>
>(({ className, ...props }, ref) => (
  <h2
    ref={ref}
    className={cn(
      "text-lg font-semibold leading-none tracking-tight",
      className,
    )}
    {...props}
  />
));

DialogTitle.displayName = "DialogTitle";
