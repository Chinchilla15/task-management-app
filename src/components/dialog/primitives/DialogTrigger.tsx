import { Children, cloneElement } from "react";
import { Button } from "@/components/common/Button";
import { useDialogContext } from "../context/DialogContext";
import type { DialogTriggerProps } from "@types";

export const DialogTrigger = ({ asChild, children }: DialogTriggerProps) => {
  const { onOpenChange } = useDialogContext();

  const child = asChild ? Children.only(children) : <Button>{children}</Button>;

  return cloneElement(child as React.ReactElement, {
    onClick: () => onOpenChange(true),
  });
};
