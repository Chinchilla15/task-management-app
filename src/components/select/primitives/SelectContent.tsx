import { cn } from "@/lib/utils";
import { useSelectKeyboard } from "@/hooks/useSelectKeyboard";
import { useSelectContext } from "../context/SelectContext";
import { SelectContentProps } from "@types";
import {
  Children,
  forwardRef,
  isValidElement,
  ReactElement,
  KeyboardEvent,
  cloneElement,
} from "react";

export const SelectContent = forwardRef<HTMLDivElement, SelectContentProps>(
  ({ className, children, title, ...props }, ref) => {
    const context = useSelectContext();

    const options = Children.toArray(children)
      .filter((child): child is ReactElement => isValidElement(child))
      .map((child) => ({ value: child.props.value }));

    const { focusedIndex, listRef } = useSelectKeyboard(
      context.open,
      options,
      context.value,
      (value) => {
        context.onValueChange(value);
        context.setOpen(false);
      },
      () => context.setOpen(false),
    );

    const handleKeyDown = (e: KeyboardEvent) => {
      if (e.key === "Escape") {
        e.preventDefault();
        e.stopPropagation();
        context.setOpen(false);
      }
    };

    if (!context.open) return null;

    return (
      <div
        ref={ref}
        id={context.listId}
        role="presentation"
        className={cn(
          "absolute top-full z-50 mt-1 min-w-max rounded-lg border border-neutral-2 bg-neutral-3 py-2 shadow-lg",
          className,
        )}
        onKeyDown={handleKeyDown}
        {...props}
      >
        {title && (
          <div
            className="px-4 text-body-xl font-bold text-neutral-2"
            id={`${context.listId}-title`}
          >
            {title}
          </div>
        )}
        <ul
          ref={listRef}
          role="listbox"
          aria-labelledby={title ? `${context.listId}-title` : undefined}
          tabIndex={-1}
          className="outline-none"
        >
          {Children.map(children, (child, index) => {
            if (!isValidElement(child)) return null;
            return cloneElement(child, {
              ...child.props,
              "aria-selected": child.props.value === context.value,
              "data-focused": index === focusedIndex,
              tabIndex: -1,
            });
          })}
        </ul>
      </div>
    );
  },
);

SelectContent.displayName = "SelectContent";
