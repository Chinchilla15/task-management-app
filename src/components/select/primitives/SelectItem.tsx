import { cn } from "@/lib/utils";
import { useSelectContext } from "../context/SelectContext";
import { SelectItemProps } from "@types";
import { forwardRef } from "react";

export const SelectItem = forwardRef<HTMLLIElement, SelectItemProps>(
  ({ className, children, value, ...props }, ref) => {
    const context = useSelectContext();
    const isSelected = context.value === value;

    return (
      <li
        ref={ref}
        role="option"
        aria-selected={isSelected}
        className={cn(
          "flex cursor-pointer items-center gap-2 px-4 py-1.5 text-body-m font-normal outline-none",
          "hover:bg-neutral-4 focus:bg-neutral-4",
          isSelected && "bg-neutral-4",
          (props as { "data-focused"?: boolean })["data-focused"] &&
            "bg-neutral-4",
          className,
        )}
        onClick={(e) => {
          e.stopPropagation();
          context.onValueChange(value);
          context.setOpen(false);
        }}
        {...props}
      >
        {children}
      </li>
    );
  },
);

SelectItem.displayName = "SelectItem";
