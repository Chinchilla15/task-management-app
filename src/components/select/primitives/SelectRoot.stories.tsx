import type { Meta, StoryObj } from "@storybook/react";
import { SelectRoot, SelectTrigger, SelectContent, SelectItem } from "../index";
import { useState } from "react";

const meta: Meta<typeof SelectRoot> = {
  title: "Components/SelectRoot",
  component: SelectRoot,
  parameters: {
    layout: "centered",
  },
};

export default meta;

type Story = StoryObj<typeof SelectRoot>;

export const Default: Story = {
  render: () => {
    const [value, setValue] = useState("");
    return (
      <SelectRoot
        value={value}
        onValueChange={(val) => setValue(val.toString())}
      >
        <SelectTrigger placeholder="Select an option" />
        <SelectContent>
          <SelectItem value="option1">Option 1</SelectItem>
          <SelectItem value="option2">Option 2</SelectItem>
          <SelectItem value="option3">Option 3</SelectItem>
        </SelectContent>
      </SelectRoot>
    );
  },
};
