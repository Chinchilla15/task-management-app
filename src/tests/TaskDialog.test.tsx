import { render, screen } from "@testing-library/react";
import { DialogRoot } from "@/components/dialog/primitives/DialogRoot";
import { DialogContent } from "@/components/dialog/primitives/DialogContent";
import { DialogTitle } from "@/components/dialog/primitives/DialogTitle";
import { DialogDescription } from "@/components/dialog/primitives/DialogDescription";
import { DialogHeader } from "@/components/dialog/primitives/DialogHeader";
import { DialogFooter } from "@/components/dialog/primitives/DialogFooter";

describe("Dialog Primitives", () => {
  test("Dialog content is rendered correctly", () => {
    render(
      <DialogRoot open={true} onOpenChange={() => {}}>
        <DialogContent>
          <DialogHeader>
            <DialogTitle>Dialog Title</DialogTitle>
            <DialogDescription>Dialog Description</DialogDescription>
          </DialogHeader>
          <DialogFooter>
            <button>Close</button>
          </DialogFooter>
        </DialogContent>
      </DialogRoot>,
    );

    expect(screen.getByText("Dialog Title")).toBeInTheDocument();
    expect(screen.getByText("Dialog Description")).toBeInTheDocument();
    expect(screen.getByText("Close")).toBeInTheDocument();
  });
});
