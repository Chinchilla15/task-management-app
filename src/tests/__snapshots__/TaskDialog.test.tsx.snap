// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dialog Primitives > Dialog content is rendered correctly 1`] = `<DocumentFragment />`;

exports[`Dialog Primitives > Dialog opens and closes correctly 1`] = `
<DocumentFragment>
  <button
    class="rounded-lg font-medium focus:outline-none p-2 transition-all ease-out duration-300 bg-primary-4 text-white hover:bg-primary-2 active:bg-primary-3 text-base"
  >
    Open Dialog
  </button>
</DocumentFragment>
`;
